src/main/java/javaslang/collection/HashArrayMappedTrie.java

```diff
@@ -212,80 +212,185 @@ private Object readResolve() {
+        abstract AbstractNode<K, V> removeElement(K key);
+        abstract int hash();
+        abstract K key();
+        abstract V value();
 
-        private final int hash;
-        private final int size;
-        private final List<Tuple2<K, V>> entries;
+        static <K, V> AbstractNode<K, V> mergeLeaves(int shift, LeafNode<K, V> leaf1, LeafSingleton<K, V> leaf2) {
+            final int h1 = leaf1.hash();
+            final int h2 = leaf2.hash();
+            if (h1 == h2) {
+                return new LeafList<>(h1, leaf2.key(), leaf2.value(), leaf1);
+            }
+            final int subH1 = hashFragment(shift, h1);
+            final int subH2 = hashFragment(shift, h2);
+            final int newBitmap = toBitmap(subH1) | toBitmap(subH2);
+            if (subH1 == subH2) {
+                AbstractNode<K, V> newLeaves = mergeLeaves(shift + SIZE, leaf1, leaf2);
+                return new IndexedNode<>(newBitmap, newLeaves.size(), List.of(newLeaves));
+            } else {
+                return new IndexedNode<>(newBitmap, leaf1.size() + leaf2.size(),
+                        subH1 < subH2 ? List.ofAll(leaf1, leaf2) : List.ofAll(leaf2, leaf1));
+            }
+        }
 
-        LeafNode(int hash, K key, V value) {
-            this(hash, 1, List.of(Tuple.of(key, value)));
+        @Override
+        boolean isLeaf() {
+            return true;
         }
 
-        private LeafNode(int hash, int size, List<Tuple2<K, V>> entries) {
-            this.hash = hash;
-            this.size = size;
-            this.entries = entries;
+        @Override
+        public boolean isEmpty() {
+            return false;
         }
+    }
 
-        private AbstractNode<K, V> update(K key, Option<V> value) {
-            List<Tuple2<K, V>> filtered = entries.removeFirst(t -> t._1.equals(key));
-            if (value.isEmpty()) {
-                return filtered.isEmpty() ? EmptyNode.instance() : new LeafNode<>(hash, filtered.length(), filtered);
-            } else {
-                return new LeafNode<>(hash, filtered.length() + 1, filtered.prepend(Tuple.of(key, value.get())));
-            }
+    /**
+     * Representation of a HAMT leaf node with single element.
+     *
+     * @param <K> Key type
+     * @param <V> Value type
+     */
+    final class LeafSingleton<K, V> extends LeafNode<K, V> implements Serializable {
+
+        private static final long serialVersionUID = 1L;
+
+        private final int hash;
+        private final K key;
+        private final V value;
+        private final Lazy<Integer> hashCode;
+
+        LeafSingleton(int hash, K key, V value) {
+            this.hash = hash;
+            this.key = key;
+            this.value = value;
+            this.hashCode = Lazy.of(() -> Objects.hash(key, value));
         }
 
         @Override
         Option<V> lookup(int shift, K key) {
-            if (hash != key.hashCode()) {
+            if (key.equals(this.key)) {
```

Comment:
Yes, tests will come. I'm on s.th. and am busy for a few days until I can fully focus on the tests...

the `equals` calls need to be changed all over the HAMT. we also compare `tuple._1.equals(key)` in LeafList etc.